File: SenderApp/wwwroot/history.js
(function (root) {
  'use strict';

  const safeGetItem = (storage, key) => {
    if (!storage || !key) return null;
    try {
      return storage.getItem(key);
    } catch {
      return null;
    }
  };

  const safeSetItem = (storage, key, value) => {
    if (!storage || !key) return;
    try {
      storage.setItem(key, value);
    } catch {
      // ignore storage failures
    }
  };

  const parseItems = (raw) => {
    if (!raw) return [];
    try {
      const parsed = JSON.parse(raw);
      if (!Array.isArray(parsed)) return [];
      
      // Support both old string format and new object format
      return parsed.map((item) => {
        if (typeof item === 'string') {
          // Migrate old string format to new object format
          return { text: item, timestamp: Date.now() };
        }
        if (item && typeof item === 'object' && item.text) {
          // Already in new format
          return { text: item.text, timestamp: item.timestamp || Date.now() };
        }
        return null;
      }).filter(item => item !== null);
    } catch {
      return [];
    }
  };

  const formatHistoryPreview = (item) => {
    if (typeof item === 'string') {
      return item; // fallback for old format
    }
    const text = item.text || '';
    const preview = text.length > 30 ? text.substring(0, 30) + '…' : text;
    
    let timeStr = '';
    if (item.timestamp) {
      const date = new Date(item.timestamp);
      const hours = String(date.getHours()).padStart(2, '0');
      const minutes = String(date.getMinutes()).padStart(2, '0');
      const seconds = String(date.getSeconds()).padStart(2, '0');
      timeStr = `${hours}:${minutes}:${seconds} | `;
    }
    
    return timeStr + preview;
  };

  const readHistory = (storage, historyKey) => {
    return parseItems(safeGetItem(storage, historyKey));
  };

  const writeHistory = (storage, historyKey, items) => {
    safeSetItem(storage, historyKey, JSON.stringify(items));
  };

  const clampIndex = (index, maxIndex) => {
    if (maxIndex < 0) return 0;
    if (Number.isNaN(index)) return maxIndex;
    return Math.min(Math.max(index, 0), maxIndex);
  };

  const readHistoryIndex = (storage, indexKey, maxIndex) => {
    if (maxIndex < 0) return 0;
    const raw = safeGetItem(storage, indexKey);
    if (raw === null || raw === undefined || raw === '') return maxIndex;
    const parsed = Number(raw);
    return clampIndex(parsed, maxIndex);
  };

  const writeHistoryIndex = (storage, indexKey, index) => {
    safeSetItem(storage, indexKey, String(index));
  };

  const loadHistoryState = (storage, historyKey, indexKey) => {
    const items = readHistory(storage, historyKey);
    if (items.length === 0) {
      return { items, index: 0 };
    }
    const maxIndex = items.length - 1;
    const index = readHistoryIndex(storage, indexKey, maxIndex);
    return { items, index };
  };

  const addHistoryEntry = (storage, historyKey, indexKey, text) => {
    const trimmed = (text || '').trim();
    if (!trimmed) {
      return loadHistoryState(storage, historyKey, indexKey);
    }

    const items = readHistory(storage, historyKey);
    const lastItem = items.length > 0 ? items[items.length - 1] : null;
    const lastText = lastItem ? (lastItem.text || lastItem) : '';
    
    if (items.length === 0 || lastText !== trimmed) {
      items.push({ text: trimmed, timestamp: Date.now() });
      writeHistory(storage, historyKey, items);
    }

    const index = items.length > 0 ? items.length - 1 : 0;
    if (items.length > 0) {
      writeHistoryIndex(storage, indexKey, index);
    }
    return { items, index };
  };

  root.LinuxKeyHistory = {
    readHistory,
    writeHistory,
    readHistoryIndex,
    writeHistoryIndex,
    loadHistoryState,
    addHistoryEntry,
    clampIndex,
    formatHistoryPreview,
  };
})(typeof window !== 'undefined' ? window : globalThis);

File: SenderApp/Client/src/History.fs
module SenderApp.Client.History

open System
open Fable.Core
open Fable.Core.JsInterop
open SenderApp.Client.HistoryCore

// ============================================================================
// JS INTEROP
// ============================================================================

[<Emit("window.localStorage")>]
let localStorage: obj = jsNative

[<Emit("typeof $0")>]
let jsTypeof (value: obj): string = jsNative

// ============================================================================
// STORAGE HELPERS
// ============================================================================

module private StorageOps =
    let tryGetItem (key: string) : string option =
        if String.IsNullOrWhiteSpace key then
            None
        else
            try
                let value: obj = localStorage?getItem(key)
                if isNull value then None else Some (string value)
            with _ -> None

    let trySetItem (key: string) (value: string) : unit =
        if String.IsNullOrWhiteSpace key then
            ()
        else
            try
                localStorage?setItem(key, value) |> ignore
            with _ -> ()

// ============================================================================
// HISTORY LOGIC
// ============================================================================

let historyKey = "linuxkey-history"
let historyIndexKey = "linuxkey-history-index"

let private parseItems (raw: string option) : HistoryItem list =
    match raw with
    | None -> []
    | Some json ->
        try
            let parsed: obj = Fable.Core.JS.JSON.parse json
            let items: obj[] = unbox parsed
            items
            |> Microsoft.FSharp.Collections.Array.toList
            |> List.choose (fun item ->
                if isNull item then
                    None
                elif jsTypeof item = "string" then
                    Some { text = string item; timestamp = Some (Fable.Core.JS.Constructors.Date.now()) }
                else
                    let textObj: obj = item?text
                    if isNull textObj then
                        None
                    else
                        let text = string textObj
                        let tsObj: obj = item?timestamp
                        let timestamp =
                            if isNull tsObj then None
                            else
                                try Some (unbox<float> tsObj) with _ -> None
                        Some { text = text; timestamp = timestamp })
        with _ -> []

let private toJsItem (item: HistoryItem) : obj =
    let ts: obj =
        match item.timestamp with
        | None -> null
        | Some value -> box value
    createObj [ "text" ==> item.text; "timestamp" ==> ts ]

let readHistory () : HistoryItem list =
    StorageOps.tryGetItem historyKey
    |> parseItems

let writeHistory (items: HistoryItem list) : unit =
    let jsItems = items |> List.map toJsItem |> List.toArray
    try
        let json = Fable.Core.JS.JSON.stringify jsItems
        StorageOps.trySetItem historyKey json
    with _ -> ()

let readHistoryIndex (maxIndex: int) : int =
    if maxIndex < 0 then 0
    else
        match StorageOps.tryGetItem historyIndexKey with
        | None | Some "" -> maxIndex
        | Some raw ->
            try
                clampIndex (int raw) maxIndex
            with _ -> clampIndex 0 maxIndex

let writeHistoryIndex (index: int) : unit =
    StorageOps.trySetItem historyIndexKey (string index)

let loadHistoryState () : HistoryState =
    let items = readHistory ()
    if items.IsEmpty then
        { items = []; index = 0 }
    else
        let maxIndex = items.Length - 1
        let index = readHistoryIndex maxIndex
        { items = items; index = index }

let addHistoryEntry (text: string) : HistoryState =
    match normalizeText text with
    | None -> loadHistoryState ()
    | Some trimmed ->
        let items = readHistory ()
        let state = addEntry (fun () -> Fable.Core.JS.Constructors.Date.now()) items trimmed
        if state.items <> items then
            writeHistory state.items
        writeHistoryIndex state.index
        state

let formatHistoryPreview (item: HistoryItem) : string =
    let preview =
        if item.text.Length > 30 then
            item.text.Substring(0, 30) + "…"
        else
            item.text

    match item.timestamp with
    | None -> preview
    | Some ts ->
        try
            let date: obj = Fable.Core.JS.Constructors.Date.Create(ts)
            let hours = int (date?getHours())
            let minutes = int (date?getMinutes())
            let seconds = int (date?getSeconds())
            sprintf "%02d:%02d:%02d | %s" hours minutes seconds preview
        with _ -> preview

// ============================================================================
// GLOBAL EXPORT (OPTIONAL COMPATIBILITY)
// ============================================================================

[<Global>]
let LinuxKeyHistory : obj =
    jsOptions (fun o ->
        o?readHistory <- fun () -> readHistory ()
        o?writeHistory <- fun (items: HistoryItem list) -> writeHistory items
        o?readHistoryIndex <- fun (maxIndex: int) -> readHistoryIndex maxIndex
        o?writeHistoryIndex <- fun (index: int) -> writeHistoryIndex index
        o?loadHistoryState <- fun () -> loadHistoryState ()
        o?addHistoryEntry <- fun (text: string) -> addHistoryEntry text
        o?clampIndex <- fun (index: int) (maxIndex: int) -> clampIndex index maxIndex
        o?formatHistoryPreview <- fun (item: HistoryItem) -> formatHistoryPreview item)
